### DISCUSSION TITLE
Error make test in latest docker build with cpu backend

### DISCUSSION THREAD

**Discussion by ghost** [flashlight/flashlight#508] [Q&A]
Dear Facebook research team,

I'm trying to build the flashlight cpu backend from docker and when I ran 'make test' command into the container I received this message as below:

root@aecae3b389ff:~/flashlight/build# make test

33% tests passed, 219 tests failed out of 328

The following tests FAILED:
 40 - AutogradTest.AfRefCountBasic (ILLEGAL)
 41 - AutogradTest.AfRefCountModify (ILLEGAL)
 42 - AutogradTest.AfRefCountGradient (ILLEGAL)
 43 - AutogradTest.AfArrBackwardNoMutate (ILLEGAL)
 44 - AutogradTest.Multiply (ILLEGAL)
 45 - AutogradTest.BasicOps (ILLEGAL)
 46 - AutogradTest.OperatorParenthesis (ILLEGAL)
 47 - AutogradTest.MultiplyAdd (ILLEGAL)
 48 - AutogradTest.AutogradOperatorTypeCompatibility (ILLEGAL)
 49 - AutogradTest.CastingAs (ILLEGAL)
 50 - AutogradTest.CastingAsInPlace (ILLEGAL)
 51 - AutogradTest.CastingAsDifferentGradTypes (ILLEGAL)
 52 - AutogradTest.CastingAsGrad (ILLEGAL)
 53 - AutogradTest.NoCalcGrad (ILLEGAL)
 54 - AutogradTest.MultiplySub (ILLEGAL)
 55 - AutogradTest.DivideAdd (ILLEGAL)
 56 - AutogradTest.MultiplyAddScalar (ILLEGAL)
 57 - AutogradTest.Exp (ILLEGAL)
 58 - AutogradTest.Pow (ILLEGAL)
 59 - AutogradTest.Sigmoid (ILLEGAL)
 60 - AutogradTest.Tanh (ILLEGAL)
 61 - AutogradTest.Concatenate (ILLEGAL)
 62 - AutogradTest.Split (ILLEGAL)
 63 - AutogradTest.TileAs (ILLEGAL)
 64 - AutogradTestF16.TileAsF16 (ILLEGAL)
 65 - AutogradTest.TileAs2 (ILLEGAL)
 66 - AutogradTest.Tile (ILLEGAL)
 67 - AutogradTest.Clamp (ILLEGAL)
 68 - AutogradTest.SumAs (ILLEGAL)
 69 - AutogradTest.SumAs2 (ILLEGAL)
 70 - AutogradTest.Sum (ILLEGAL)
 71 - AutogradTest.Log1p (ILLEGAL)
 72 - AutogradTest.Sqrt (ILLEGAL)
 73 - AutogradTest.Mean (ILLEGAL)
 74 - AutogradTest.Variance (ILLEGAL)
 75 - AutogradTest.Norm (ILLEGAL)
 76 - AutogradTest.Normalize (ILLEGAL)
 77 - AutogradTest.Indexing (ILLEGAL)
 78 - AutogradTest.Convolve (ILLEGAL)
 79 - AutogradTestF16.ConvolveF16 (ILLEGAL)
 80 - AutogradTest.ConvolveFilterGroups (ILLEGAL)
 81 - AutogradTest.ConvolveDilation (ILLEGAL)
 82 - AutogradTest.Padding (ILLEGAL)
 83 - AutogradTest.Pooling (ILLEGAL)
 84 - AutogradTestF16.PoolingF16 (ILLEGAL)
 85 - AutogradTest.Softmax (ILLEGAL)
 86 - AutogradTestF16.SoftmaxF16 (ILLEGAL)
 87 - AutogradTest.LogSoftmax (ILLEGAL)
 88 - AutogradTestF16.LogSoftmaxF16 (ILLEGAL)
 89 - AutogradTest.BinaryCrossEntropy (ILLEGAL)
 90 - AutogradTest.CrossEntropy (ILLEGAL)
 91 - AutogradTest.Reorder (ILLEGAL)
 92 - AutogradTest.Glu (ILLEGAL)
 93 - AutogradTest.Linear (ILLEGAL)
 94 - AutogradTestF16.LinearF16 (ILLEGAL)
 95 - AutogradTest.WeightNormLinear (ILLEGAL)
 96 - AutogradTest.WeightNormConv (ILLEGAL)
 97 - AutogradTest.Rnn (ILLEGAL)
 98 - AutogradTestF16.RnnF16 (ILLEGAL)
 99 - AutogradTest.Lstm (ILLEGAL)
100 - AutogradTestF16.LstmF16 (ILLEGAL)
101 - AutogradTest.Gru (ILLEGAL)
102 - AutogradTestF16.GruF16 (ILLEGAL)
103 - AutogradTest.Embedding (ILLEGAL)
104 - AutogradTest.BatchNormEvalModeOutputSingleAxis (ILLEGAL)
105 - AutogradTest.BatchNormEvalModeOutputMultipleAxis (ILLEGAL)
106 - AutogradTest.BatchNormTrainModeOutputSingleAxis (ILLEGAL)
107 - AutogradTest.BatchNormTrainModeOutputMultipleAxis (ILLEGAL)
108 - AutogradTest.BatchNormJacobian (ILLEGAL)
109 - AutogradTestF16.BatchNormJacobianF16 (ILLEGAL)
110 - AutogradTest.BatchNormJacobianMultipleAxes (ILLEGAL)
111 - AutogradTestF16.BatchNormJacobianMultipleAxesF16 (ILLEGAL)
112 - AutogradTest.LayerNormJacobian (ILLEGAL)
113 - AutogradTestF16.LayerNormJacobianF16 (ILLEGAL)
114 - AutogradTest.GetAdvancedIndex (ILLEGAL)
115 - AutogradTest.GetAdvancedIndexF16 (ILLEGAL)
148 - ModuleTest.EmbeddingFwd (ILLEGAL)
149 - ModuleTest.LinearFwd (ILLEGAL)
150 - ModuleTestF16.LinearFwdF16 (ILLEGAL)
151 - ModuleTest.ConvPadding (ILLEGAL)
152 - ModuleTest.GLUFwd (ILLEGAL)
153 - ModuleTestF16.GLUFwdF16 (ILLEGAL)
154 - ModuleTest.LogSoftmaxFwd (ILLEGAL)
155 - ModuleTestF16.LogSoftmaxFwdF16 (ILLEGAL)
156 - ModuleTest.ConvolutionFwd (ILLEGAL)
157 - ModuleTestF16.ConvolutionFwdF16 (ILLEGAL)
158 - ModuleTest.ConvolutionWithGroupFwd (ILLEGAL)
159 - ModuleTest.PoolingFwd (ILLEGAL)
160 - ModuleTestF16.PoolingFwdF16 (ILLEGAL)
161 - ModuleTest.RNNFwd (ILLEGAL)
162 - ModuleTest.LSTMFwd (ILLEGAL)
163 - ModuleTest.GRUFwd (ILLEGAL)
164 - ModuleTestF16.RNNFwdF16 (ILLEGAL)
165 - ModuleTest.ViewFwd (ILLEGAL)
166 - ModuleTest.DropoutFwd (ILLEGAL)
167 - ModuleTestF16.DropoutFwdF16 (ILLEGAL)
168 - ModuleTest.PaddingFwd (ILLEGAL)
169 - ModuleTest.LayerNormFwd (ILLEGAL)
170 - ModuleTestF16.LayerNormFwdF16 (ILLEGAL)
171 - ModuleTest.NormalizeFwd (ILLEGAL)
172 - ModuleTest.TransformFwd (ILLEGAL)
173 - ModuleTest.PrecisionCastFwd (ILLEGAL)
174 - ModuleTest.ContainerReplaceParam (ILLEGAL)
175 - ModuleTest.AdaptiveSoftMaxPredict (ILLEGAL)
176 - ModuleTest.AdaptiveSoftMaxLossBatchFwd (ILLEGAL)
177 - ModuleTest.AdaptiveSoftMaxLossIgnoreIndex (ILLEGAL)
178 - ModuleTest.IdentityFwd (ILLEGAL)
179 - NNSerializationTest.Variable (ILLEGAL)
180 - NNSerializationTest.Linear (ILLEGAL)
181 - NNSerializationTest.Conv2D (ILLEGAL)
182 - NNSerializationTest.Pool2D (ILLEGAL)
183 - NNSerializationTest.BaseModule (ILLEGAL)
184 - NNSerializationTest.PrecisionCast (ILLEGAL)
185 - NNSerializationTest.WeightNormLinear (ILLEGAL)
186 - NNSerializationTest.WeightNormConvSeq (ILLEGAL)
187 - NNSerializationTest.AdaptiveSoftMaxLoss (ILLEGAL)
188 - NNSerializationTest.PrettyString (ILLEGAL)
189 - NNSerializationTest.LeNet (ILLEGAL)
190 - NNSerializationTest.FileSize (ILLEGAL)
191 - NNSerializationTest.VariableTwice (ILLEGAL)
192 - NNSerializationTest.ContainerBackward (ILLEGAL)
193 - NNSerializationTest.ContainerWithParams (ILLEGAL)
194 - UtilsTest.Join (ILLEGAL)
220 - ContribModuleTest.ResidualFwd (ILLEGAL)
221 - ContribModuleTest.ResidualFwdWithProjection (ILLEGAL)
222 - ContribModuleTest.AsymmetricConv1DFwd (ILLEGAL)
223 - ContribModuleTest.TransformerPadMaskFwd (ILLEGAL)
224 - ContribModuleTestF16.TransformerPadMaskFwd16 (ILLEGAL)
225 - ContribModuleTest.TransformerFwd (ILLEGAL)
226 - ContribModuleTestF16.TransformerFwdF16 (ILLEGAL)
227 - ContribModuleTest.ConformerFwd (ILLEGAL)
228 - ContribModuleTestF16.ConformerFwdF16 (ILLEGAL)
229 - ContribModuleTest.PositionEmbeddingFwd (ILLEGAL)
230 - ContribModuleTestF16.PositionEmbeddingFwdF16 (ILLEGAL)
231 - ContribModuleTest.SinusoidalPositionEmbeddingFwd (ILLEGAL)
232 - ContribModuleTestF16.SinusoidalPositionEmbeddingFwdF16 (ILLEGAL)
233 - ContribModuleTest.AdaptiveEmbedding (ILLEGAL)
234 - ContribModuleTest.TDSFwd (ILLEGAL)
235 - ContribModuleTest.StreamingTDSFwd (ILLEGAL)
236 - ContribModuleTest.SpecAugmentFwd (ILLEGAL)
237 - ContribModuleTest.RawWavSpecAugmentFwd (ILLEGAL)
238 - SerializationTest.Residual (ILLEGAL)
239 - SerializationTest.AsymmetricConv1D (ILLEGAL)
240 - SerializationTest.Transformer (ILLEGAL)
241 - SerializationTest.ConformerSerialization (ILLEGAL)
242 - SerializationTest.PositionEmbedding (ILLEGAL)
243 - SerializationTest.SinusoidalPositionEmbedding (ILLEGAL)
244 - SerializationTest.AdaptiveEmbedding (ILLEGAL)
245 - SerializationTest.RawWavSpecAugment (ILLEGAL)
246 - SequentialBuilderTest.SeqModule (ILLEGAL)
247 - SequentialBuilderTest.Serialization (ILLEGAL)
248 - ModulePluginTest.ModulePlugin (ILLEGAL)
249 - CriterionTest.CTCEmptyTarget (ILLEGAL)
250 - CriterionTest.CTCCost (ILLEGAL)
251 - CriterionTest.CTCJacobian (ILLEGAL)
252 - CriterionTest.Batching (ILLEGAL)
253 - CriterionTest.CTCCompareTensorflow (ILLEGAL)
254 - CriterionTest.ViterbiPath (ILLEGAL)
255 - CriterionTest.ASGAlternatingBlanks (ILLEGAL)
256 - CriterionTest.ViterbiPathConstrained (ILLEGAL)
257 - CriterionTest.CTCViterbiPathNopaces (ILLEGAL)
258 - CriterionTest.CTCViterbiPathConstrainedEndWithSpace (ILLEGAL)
259 - CriterionTest.CTCViterbiPathConstrainedBeginWithSpace (ILLEGAL)
260 - CriterionTest.CTCViterbiPathConstrainedBeginAndEndWithSpace (ILLEGAL)
261 - CriterionTest.FCCCost (ILLEGAL)
262 - CriterionTest.FCCJacobian (ILLEGAL)
263 - CriterionTest.FACCost (ILLEGAL)
264 - CriterionTest.FACJacobian (ILLEGAL)
265 - CriterionTest.ASGCost (ILLEGAL)
266 - CriterionTest.ASGJacobian (ILLEGAL)
267 - CriterionTest.LinSegJacobian (ILLEGAL)
268 - CriterionTest.ASGBatching (ILLEGAL)
269 - CriterionTest.ASGCompareLua (ILLEGAL)
270 - CriterionTest.LinSegCompareLua (ILLEGAL)
271 - CriterionTest.AsgSerialization (ILLEGAL)
272 - Seq2SeqTest.Seq2Seq (ILLEGAL)
273 - Seq2SeqTest.Seq2SeqViterbi (ILLEGAL)
274 - Seq2SeqTest.Seq2SeqBeamSearchViterbi (ILLEGAL)
275 - Seq2SeqTest.Seq2SeqMedianWindow (ILLEGAL)
276 - Seq2SeqTest.Seq2SeqStepWindow (ILLEGAL)
277 - Seq2SeqTest.Seq2SeqStepWindowVectorized (ILLEGAL)
278 - Seq2SeqTest.Seq2SeqAttn (ILLEGAL)
279 - Seq2SeqTest.Seq2SeqMixedAttn (ILLEGAL)
280 - Seq2SeqTest.Serialization (ILLEGAL)
281 - Seq2SeqTest.BatchedDecoderStep (ILLEGAL)
282 - Seq2SeqTest.Seq2SeqSampling (ILLEGAL)
283 - AttentionTest.NeuralContentAttention (ILLEGAL)
284 - AttentionTest.SimpleLocationAttention (ILLEGAL)
285 - AttentionTest.LocationAttention (ILLEGAL)
286 - AttentionTest.NeuralLocationAttention (ILLEGAL)
287 - AttentionTest.MultiHeadContentAttention (ILLEGAL)
288 - AttentionTest.JacobianMaskAttention (ILLEGAL)
289 - WindowTest.MedianWindow (ILLEGAL)
290 - WindowTest.MedianWindowWithPad (ILLEGAL)
291 - WindowTest.StepWindow (ILLEGAL)
292 - WindowTest.SoftWindow (ILLEGAL)
293 - WindowTest.SoftPretrainWindow (ILLEGAL)
294 - FeaturizationTest.AfMatmulCompare (ILLEGAL)
295 - FeaturizationTest.Normalize (ILLEGAL)
296 - FeaturizationTest.Transpose (ILLEGAL)
297 - FeaturizationTest.localNormalize (ILLEGAL)
298 - FeaturizationTest.TargetTknTestStandaloneSep (ILLEGAL)
299 - FeaturizationTest.TargetTknTestInsideSep (ILLEGAL)
300 - FeaturizationTest.WrdToTarget (ILLEGAL)
301 - FeaturizationTest.TargetToSingleLtr (ILLEGAL)
302 - FeaturizationTest.inputFeaturizer (ILLEGAL)
303 - FeaturizationTest.targetFeaturizer (ILLEGAL)
304 - ListFileDatasetTest.LoadData (ILLEGAL)
309 - ConvLmModuleTest.GCNN14BAdaptiveSoftmax (ILLEGAL)
310 - ConvLmModuleTest.GCNN14BCrossEntropy (ILLEGAL)
311 - ConvLmModuleTest.SerializationGCNN14BAdaptiveSoftmax (ILLEGAL)
312 - DecoderTest.run (ILLEGAL)
313 - RuntimeTest.LoadAndSave (ILLEGAL)
314 - RuntimeTest.TestCleanFilepath (ILLEGAL)
315 - RuntimeTest.SpeechStatMeter (ILLEGAL)
316 - RuntimeTest.parseValidSets (ILLEGAL)
327 - TextDatasetTest.NoneMode (ILLEGAL)
328 - TextDatasetTest.EosMode (ILLEGAL)
329 - TextDatasetTest.EosModeWithDynamicBatching (ILLEGAL)
Errors while running CTest
make: *** [Makefile:107: test] Error 8

Thanks for your help!

**Comment by tlikhomanenko** [comment]
Can you point to the image you tried? did you build it by yourself? Can you run one of these tests directly to post what exactly the error there?

**Reply by ghost** [reply]
Hi Tatiana,

I ran flashlight with docker image for cpu backend as in https://github.com/facebookresearch/flashlight/tree/master/.docker

and when I ran AutogradTest I received this message as below:

root@eef782a261fc:~/flashlight/build/flashlight/fl/test# ./AutogradTest
Illegal instruction (core dumped)

**Comment by ghost** [comment]
I would like to know if there were any minimum requirements on processor, RAM and OS ubuntu version to use the flashlight CPU backend from Docker.

I tried to run the "make test" command on the flashlight CPU backend from Docker using different virtual environments with different processors, RAM and ubuntu operating systems.

Below I describe the different configurations and the results obtained:

>>>>>> 1 <<<<<<
Operating system: Ubuntu 18.04.5 LTS x86_64
CPU: Intel Xeon E7-4830 v2 (8) at 2.194 GHz
RAM: 32GB
make test ---> 100% pass tests, 0 tests fail out of 328
and the asr service works fine!

... but if I reduce the RAM to 16GB
make test ---> 33% of tests passed, 219 tests failed out of 328
the message shown when I run a single failed test is
"Illegal education (core dumped)"
and the asr service doesn't work!

>>>>>> 2 <<<<<<
Operating system: Ubuntu 20.04.2 LTS x86_64
CPU: Intel Xeon E7-4830 v2 (16) at 2.194 GHz
GPU: 00: 02.0 Cirrus Logic GD 5446
RAM: 32GB

make test ---> 99% of tests passed, 4 out of 371 tests failed
and the asr service works fine!

>>>>>> 3 <<<<<<
Operating system: Ubuntu 20.04.2 LTS x86_64
PROCESSOR: AMD Opteron 6164 HE (16) at 1.695 GHz
GPU: 00: 02.0 Cirrus Logic GD 5446
RAM: 64GB

make test ---> 35% of tests passed, 240 out of 371 tests failed
the message shown when I run a single failed test is
"Illegal education (core dumped)"
and the asr service doesn't work!

Therefore, I would like to know why reducing the RAM from 32GB to 16GB or changing ubuntu os from 18.04 to 20.04 version or with different processor the asr service does not work? 

Also I would like to know why changing ubuntu os from 18.04 to 20.04 version the number of tests carried out has gone from 328 to 371?

**Reply by tlikhomanenko** [reply]
Wow! I didn't expect this really, thanks for detailed debugging and report along with your system info! cc @jacobkahn @avidov do you have an idea why this happens and what do we need to fix in our tests?

**Reply by avidov** [reply]
Thank you for the detailed report.

> I would like to know if there were any minimum requirements on processor, RAM and OS ubuntu version to use the flashlight CPU backend from Docker.

CPU backend is less extensively tested and studied compare to CUDA. I do not have numbers for that.
Generally when running the tests over CUDA, memory usage is below 1GB with few spikes > 10GB.

> "Illegal education (core dumped)"

From a quick search it seems to do with AVX instruction set support. Possibly due to HW / driver mismatch.

Your benchmark results are very interesting. Did you monitor RAM usage while running the tests?
Did you see the RAM utilization go to 100%

Curious what are your goals?

If you are using Flashlight over CPU for a product over x86, and need maximum performance, I'd suggest installing Inetl OpenCL driver (by installing oneDNN) and building Flashlight with OpenCL backend.
